### PR TITLE
Optimization of SP adjustment code after forward gotos.

### PR DESCRIPTION
**Pull request by laubzega**
A fix for issue #769.

### CODE REVIEW

File: src/cc65/symtab.c [hunk 1]
```diff
@@ -677,6 +680,18 @@ DefOrRef* AddDefOrRef(SymEntry* E, unsigned Flags)
     return DOR;
 }
 
+unsigned short FindSPAdjustment (const char* Name)
+{
+/* Search for an entry in the table of SP adjustments */
```
Comment (oliverschmidt) [inline src/cc65/symtab.c]: May I ask to put the comment above the brackets?

File: src/cc65/symtab.c [hunk 2]
```diff
@@ -728,10 +743,17 @@ SymEntry* AddLabelSym (const char* Name, unsigned Flags)
                 /* We're processing a label, let's update all gotos encountered
                 ** so far
                 */
+                SymEntry *E;
                 g_userodata();
                 g_defdatalabel (DOR->LateSP_Label);
                 g_defdata (CF_CONST | CF_INT, StackPtr - DOR->StackPtr, 0);
 
+                /* Optimizer will need the information about the value of SP adjustment
+                ** later, so let's preserve it. */
```
Comment (oliverschmidt) [inline src/cc65/symtab.c]: The predominant style wants to have the comment closing of multi-line comments on its own line.

File: src/cc65/codeopt.c [hunk 3]
```diff
@@ -613,7 +613,115 @@ static unsigned OptStackPtrOps (CodeSeg* S)
...
+            L[2]->OPC == OP65_CLC            &&
+            L[3]->OPC == OP65_ADC            &&
+            strcmp (L[3]->Arg, "sp") == 0    &&
+            L[6]->OPC == OP65_ADC            &&
+            strcmp (L[6]->Arg, "sp+1") == 0  &&
+            L[9]->OPC == OP65_JMP) {
+            adjustment = FindSPAdjustment (L[1]->Arg);
+
+            if (adjustment == 0) {
+            /* No SP adjustment needed, remove the whole sequence */
```
Comment (oliverschmidt) [inline src/cc65/codeopt.c]: Please align the comment with the actual code.

File: src/cc65/codeopt.c [hunk 4]
```diff
@@ -613,7 +613,115 @@ static unsigned OptStackPtrOps (CodeSeg* S)
...
+                CS_DelEntries (S, I, 9);
+            }
+            else if (adjustment >= 65536 - 8) {
+            /* If adjustment is in range [-8, 0) we use decsp* calls */
```
Comment (oliverschmidt) [inline src/cc65/codeopt.c]: Please align the comment with the actual code.

File: src/cc65/codeopt.c [hunk 5]
```diff
@@ -613,7 +613,115 @@ static unsigned OptStackPtrOps (CodeSeg* S)
...
+                adjustment = 65536 - adjustment;
+                xsprintf (Buf, sizeof (Buf), "decsp%u", adjustment);
+                X = NewCodeEntry (OP65_JSR, AM65_ABS, Buf, 0, L[1]->LI);
+                CS_InsertEntry (S, X, I + 9);
+
+                /* Delete the old code */
+                CS_DelEntries (S, I, 9);
+            }
+            else if (adjustment >= 65536 - 255) {
+            /* For range [-255, -8) we have ldy #, jsr subysp */
```
Comment (oliverschmidt) [inline src/cc65/codeopt.c]: Please align the comment with the actual code.

File: src/cc65/codeopt.c [hunk 6]
```diff
@@ -613,7 +613,115 @@ static unsigned OptStackPtrOps (CodeSeg* S)
...
+                CS_InsertEntry (S, X, I + 9);
+                X = NewCodeEntry (OP65_JSR, AM65_ABS, "subysp", 0, L[1]->LI);
+                CS_InsertEntry (S, X, I + 10);
+
+                /* Delete the old code */
+                CS_DelEntries (S, I, 9);
+            }
+            else if (adjustment > 255) {
+            /* For ranges [-32768, 255) and (255, 32767) the only modification
+            ** is to replace the absolute with immediate addressing */
```
Comment (oliverschmidt) [inline src/cc65/codeopt.c]: Please align the comment with the actual code and have the comment closing on its own line.

File: src/cc65/codeopt.c [hunk 7]
```diff
@@ -613,7 +613,115 @@ static unsigned OptStackPtrOps (CodeSeg* S)
...
+                Arg = MakeHexArg (adjustment >> 8);
+                X = NewCodeEntry (OP65_LDA, AM65_IMM, Arg, 0, L[5]->LI);
+                CS_InsertEntry (S, X, I + 6);
+
+                /* Delete the old code */
+                CS_DelEntry (S, I + 2);
+                CS_DelEntry (S, I + 6);
+            }
+            else if (adjustment > 8) {
+            /* For range (8, 255] we have ldy #, jsr addysp */
```
Comment (oliverschmidt) [inline src/cc65/codeopt.c]: Please align the comment with the actual code.

File: src/cc65/codeopt.c [hunk 8]
```diff
@@ -613,7 +613,115 @@ static unsigned OptStackPtrOps (CodeSeg* S)
...
+                X = NewCodeEntry (OP65_LDY, AM65_IMM, Arg, 0, L[1]->LI);
+                CS_InsertEntry (S, X, I + 9);
+                X = NewCodeEntry (OP65_JSR, AM65_ABS, "addysp", 0, L[1]->LI);
+                CS_InsertEntry (S, X, I + 10);
+
+                /* Delete the old code */
+                CS_DelEntries (S, I, 9);
+            }
+            else {
+            /* If adjustment is in range (0, 8] we use incsp* calls */
```
Comment (oliverschmidt) [inline src/cc65/codeopt.c]: Please align the comment with the actual code.

File: src/cc65/symtab.c [hunk 9]
```diff
@@ -405,7 +408,7 @@ void LeaveStructLevel (void)
 
 
 
-static SymEntry* FindSymInTable (const SymTable* T, const char* Name, unsigned Hash)
+SymEntry* FindSymInTable (const SymTable* T, const char* Name, unsigned Hash)
```
Comment (greg-king5) [inline src/cc65/symtab.c]: `FindSymInTable()` should continue to be a private (static) function.

**Pull request merged**